src/com/ms/silverking/cloud/dht/client/gen/WrapperGenerator.java

```diff
@@ -359,494 +372,502 @@ private void generate(Context c, List<ParseElement> elements, File outputDir, in
+  }
+
+  private void generateLoopTargetException(LoopElement le, Target expectedTarget) {
+    throw new RuntimeException(String.format("Invalid target for %s. Expected %s.", le, expectedTarget));
+  }
+
+  /**
+   * Currently this method is only for debugging
+   *
+   * @param templateFile
+   * @param outputDir
+   * @throws IOException
+   */
+  public void generate(File templateFile, File outputDir) throws IOException {
+    if (!outputDir.exists()) {
+      throw new RuntimeException("outputDir does not exist: " + outputDir);
+    } else {
+      List<ParseElement> elements;
+
+      elements = parseElements(templateFile);
+      //displayElements(elements);
+    }
+  }
+
+  private void generate(List<ParseElement> elements, Context c) {
+    for (int i = 0; i < elements.size(); i++) {
+      ParseElement e;
+
+      e = elements.get(i);
+      if (e instanceof Text) {
+        System.out.print(e);
+      } else if (e instanceof Expression) {
+        ((Expression) e).evaluate(c);
+      } else if (e instanceof LoopElement) {
+        LoopElement le;
+
+        le = (LoopElement) e;
+        loop(le, getEnclosedStatements(elements, i), c);
+        generate(getEnclosedStatements(elements, i), c);
+        //generate(getEndingLoopElementIndex(le, elements)
+      } else {
+        throw new RuntimeException("Panic");
+      }
+    }
+  }
+
+  private void loop(LoopElement e, List<ParseElement> enclosedStatements, Context c) {
+  }
+
+  private List<ParseElement> getEnclosedStatements(List<ParseElement> elements, int i) {
+    return null;
+  }
+
+  ////////////////////////////////////////////////////////////////////
+
+  private void displayElements(List<ParseElement> elements) {
+    for (int i = 0; i < elements.size(); i++) {
+      System.out.printf("%d: %s\n\n", i, elements.get(i));
+    }
+  }
+
+  ////////////////////////////////////////////////////////////////////
+
+  private int nextDelimiter(String s, String d, int i) {
+    int next;
+
+    next = s.indexOf(d, i);
+    if (next < 0) {
+      next = s.length();
+    }
+    return next;
+  }
+
+  private void generateParseException(String s, String message, int location) {
+    int line;
+
+    line = StringUtil.countOccurrences(s, '\n', location) + 1;
+    throw new RuntimeException(String.format("Line: %d. %s", line, message));
+  }
+
+  /**
+   * Parse the template file into a list of ParseElements
+   *
+   * @param templateFile
+   * @return list of ParseElements
+   * @throws IOException
+   */
+  private List<ParseElement> parseElements(File templateFile) throws IOException {
+    String s;
+    byte[] buf;
+    int nextStart;
+    int nextEnd;
+    int i;
+    int length;
+    ParseState state;
+    List<ParseElement> elements;
+
+    elements = new ArrayList<>();
+    length = Math.toIntExact(templateFile.length());
+    buf = new byte[length];
+    StreamUtil.readFully(new FileInputStream(templateFile), buf);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *FileInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when StreamUtil.readFully throws an exception. Either a) close the object returned by *FileInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *FileInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).